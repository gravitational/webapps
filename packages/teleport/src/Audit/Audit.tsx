/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import EventList from './EventList';
import RangePicker from 'teleport/components/EventRangePicker';
import { Danger } from 'design/Alert';
import { Flex, Indicator, Box } from 'design';
import InputSearch from 'teleport/components/InputSearch';
import useTeleport from 'teleport/useTeleport';
import useStickyClusterId from 'teleport/useStickyClusterId';
import useAuditEvents from 'teleport/useAuditEvents';
import EventsFilter from './EventsFilter';

export default function Container() {
  const teleCtx = useTeleport();
  const { clusterId } = useStickyClusterId();
  const state = useAuditEvents(teleCtx, clusterId);
  return <Audit {...state} />;
}

export function Audit(props: ReturnType<typeof useAuditEvents>) {
  const {
    moreEvents,
    onFilterChange,
    attempt,
    range,
    setRange,
    rangeOptions,
    events,
    searchValue,
    clusterId,
    setSearchValue,
  } = props;

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center">
        <FeatureHeaderTitle mr="8">Audit Log</FeatureHeaderTitle>
        <Flex ml="auto">
          <EventsFilter onFilterChange={onFilterChange} />
          <RangePicker
            ml="3"
            range={range}
            ranges={rangeOptions}
            onChangeRange={setRange}
          />
        </Flex>
      </FeatureHeader>
      <Flex
        mb={4}
        alignItems="center"
        flex="0 0 auto"
        justifyContent="flex-start"
      >
        <InputSearch mr="3" onChange={setSearchValue} />
      </Flex>
      {attempt.status === 'failed' && <Danger> {attempt.statusText} </Danger>}
      {attempt.status === 'processing' && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.status === 'success' && (
        <EventList
          search={searchValue}
          events={events}
          clusterId={clusterId}
          moreEvents={moreEvents}
        />
      )}
    </FeatureBox>
  );
}
