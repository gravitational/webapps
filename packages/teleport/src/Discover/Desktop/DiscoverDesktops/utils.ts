import { Path } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/Path';
import { MonotoneX } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/MonotoneX';

const STROKE_WIDTH = 4;

export interface Line {
  width: number;
  height: number;
  path: string;
}

export function createLine(
  desktopServiceElement: HTMLDivElement,
  desktopElement: HTMLDivElement,
  containerElement: HTMLDivElement
): Line {
  if (!desktopElement || !desktopServiceElement || !containerElement) {
    return null;
  }

  const desktopServiceRect = desktopServiceElement.getBoundingClientRect();
  const desktopRect = desktopElement.getBoundingClientRect();
  const containerRect = containerElement.getBoundingClientRect();

  const distance = desktopRect.left - desktopServiceRect.right;

  const path = new Path();
  const line = new MonotoneX(path);

  line.lineStart();

  const desktopLinePosition =
    desktopRect.top - containerRect.top + desktopRect.height / 2 - 1;
  const desktopServiceLinePosition =
    desktopServiceRect.top - containerRect.top + desktopServiceRect.height / 2;

  line.point(0, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(40, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(distance - 10, desktopLinePosition + STROKE_WIDTH / 2);
  line.point(distance, desktopLinePosition + STROKE_WIDTH / 2);

  line.lineEnd();

  return {
    width: distance,
    height: containerRect.height,
    path: path.path,
  };
}
