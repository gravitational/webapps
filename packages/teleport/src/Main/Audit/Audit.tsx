/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { useEffect, useState, useMemo } from 'react';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import EventList from './EventList';
import RangePicker, { getRangeOptions } from './RangePicker';
import { useAttempt } from 'shared/hooks';
import { Danger } from 'design/Alert';
import { Flex, Indicator, Box } from 'design';
import InputSearch from 'teleport/components/InputSearch';
import TeleportContext from 'teleport/teleportContext';
import { useTeleport } from 'teleport/teleportContextProvider';
import useClusterId from 'teleport/Main/useClusterId';

export default function Container() {
  const clusterId = useClusterId();
  const teleCtx = useTeleport();
  const state = useEvents(clusterId, teleCtx);
  return <Audit {...state} />;
}

export function Audit(props: ReturnType<typeof useEvents>) {
  const {
    overflow,
    attempt,
    maxLimit,
    range,
    rangeOptions,
    setRange,
    events,
    searchValue,
    clusterId,
    setSearchValue,
  } = props;

  const { isSuccess, isFailed, message, isProcessing } = attempt;

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center">
        <FeatureHeaderTitle mr="8">Audit Log</FeatureHeaderTitle>
        <RangePicker
          ml="auto"
          value={range}
          options={rangeOptions}
          onChange={setRange}
        />
      </FeatureHeader>
      <Flex mb={4} alignItems="center" justifyContent="flex-start">
        <InputSearch height="30px" mr="3" onChange={setSearchValue} />
      </Flex>
      {overflow && (
        <Danger>
          Number of events retrieved for specified date range has exceeded the
          maximum limit of {maxLimit} events
        </Danger>
      )}
      {isFailed && <Danger> {message} </Danger>}
      {isProcessing && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {isSuccess && (
        <EventList search={searchValue} events={events} clusterId={clusterId} />
      )}
    </FeatureBox>
  );
}

function useEvents(clusterId: string, ctx: TeleportContext) {
  const rangeOptions = useMemo(() => getRangeOptions(), []);
  const [searchValue, setSearchValue] = React.useState('');
  const [range, setRange] = useState(rangeOptions[0]);
  const [attempt, attemptActions] = useAttempt({ isProcessing: true });
  const [results, setResults] = useState({
    events: [],
    overflow: false,
  });

  function onFetch({ from, to }: Range) {
    attemptActions.do(() => {
      return ctx.auditService
        .fetchEvents(clusterId, { start: from, end: to })
        .then(setResults);
    });
  }

  useEffect(() => {
    onFetch(range);
  }, [clusterId, range]);

  return {
    ...results,
    attempt,
    clusterId,
    attemptActions,
    onFetch,
    maxLimit: ctx.auditService.maxLimit,
    range,
    rangeOptions,
    setRange,
    searchValue,
    setSearchValue,
  };
}

type Range = { from: Date; to: Date };
