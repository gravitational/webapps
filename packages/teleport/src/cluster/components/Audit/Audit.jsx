/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { Redirect, Switch, Route } from 'shared/components/Router';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import Tabs, { TabItem } from './Tabs';
import AuditSessions from './AuditSessions';
import AuditEvents from './AuditEvents';
import RangePicker, { getRangeOptions } from './RangePicker';
import cfg from 'teleport/config';

export default function Audit() {
  const rangeOptions = React.useMemo(() => getRangeOptions(), []);
  const [range, handleOnRange] = React.useState(rangeOptions[0]);
  const auditRoute = cfg.getAuditRoute();
  const eventsRoute = cfg.getAuditEventsRoute();
  const sessionsRoute = cfg.getAuditSessionsRoute();

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center">
        <FeatureHeaderTitle mr="8">Audit Log</FeatureHeaderTitle>
        <Tabs>
          <TabItem to={eventsRoute} title="Events" />
          <TabItem to={sessionsRoute} title="Sessions" />
        </Tabs>
        <RangePicker
          ml="auto"
          value={range}
          options={rangeOptions}
          onChange={handleOnRange}
        />
      </FeatureHeader>
      <Switch>
        <Route title="Events" path={eventsRoute}>
          <AuditEvents range={range} />
        </Route>
        <Route
          title="Sessions"
          path={sessionsRoute}
          component={AuditSessions}
        />
        <Redirect exact from={auditRoute} to={eventsRoute} />
      </Switch>
    </FeatureBox>
  );
}
