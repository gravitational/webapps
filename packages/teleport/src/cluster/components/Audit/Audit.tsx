/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { useEffect, useState, useMemo } from 'react';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import AuditEvents from './AuditEvents';
import RangePicker, { getRangeOptions } from './RangePicker';
import { useAttempt } from 'shared/hooks';
import { Danger } from 'design/Alert';
import { Indicator, Box } from 'design';
import { useTeleport } from 'teleport/teleportContextProvider';

export default function Audit() {
  const rangeOptions = useMemo(() => getRangeOptions(), []);
  const [range, handleOnRange] = useState(rangeOptions[0]);

  const { overflow, attempt, maxLimit, events } = useEvents(range);
  const { isSuccess, isFailed, message, isProcessing } = attempt;

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center">
        <FeatureHeaderTitle mr="8">Audit Log</FeatureHeaderTitle>
        <RangePicker
          ml="auto"
          value={range}
          options={rangeOptions}
          onChange={handleOnRange}
        />
      </FeatureHeader>
      {overflow && (
        <Danger>
          Number of events retrieved for specified date range has exceeded the
          maximum limit of {maxLimit} events
        </Danger>
      )}
      {isFailed && <Danger> {message} </Danger>}
      {isProcessing && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {isSuccess && <AuditEvents events={events} />}
    </FeatureBox>
  );
}

function useEvents(range: Range) {
  const teleCtx = useTeleport();
  const [attempt, attemptActions] = useAttempt({ isProcessing: true });
  const [results, setResults] = useState({
    events: [],
    overflow: false,
  });

  function onFetch({ from, to }: Range) {
    attemptActions.do(() => {
      return teleCtx.auditService
        .fetchEvents({ start: from, end: to })
        .then(setResults);
    });
  }

  function onFetchLatest() {
    return teleCtx.auditService.fetchLatest().then(setResults);
  }

  useEffect(() => {
    onFetch(range);
  }, [range]);

  return {
    ...results,
    attempt,
    attemptActions,
    onFetch,
    onFetchLatest,
    maxLimit: teleCtx.auditService.maxLimit,
  };
}

type Range = { from: Date; to: Date };
