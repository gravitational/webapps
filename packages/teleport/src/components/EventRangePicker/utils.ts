import moment from 'moment';

export function getRangeOptions(): EventRange[] {
  return [
    {
      name: 'Today',
      from: moment(new Date())
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: '7 days',
      from: moment()
        .subtract(6, 'day')
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: 'Custom Range...',
      isCustom: true,
      from: new Date(),
      to: new Date(),
    },
  ];
}

export type EventRange = {
  from: Date;
  to: Date;
  isCustom?: boolean;
  name?: string;
};
