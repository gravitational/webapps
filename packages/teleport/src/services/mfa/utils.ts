/**
 * Copyright 2021 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Auth2faType, PreferredMfaType } from 'shared/services';

export type MfaOption = {
  value: Auth2faType;
  label: string;
};

export function getMfaOptions(
  mfa: Auth2faType,
  preferredMfa: PreferredMfaType,
  requireMfa = false
) {
  const mfaOptions: MfaOption[] = [];

  if (mfa === 'off' || !mfa) {
    return mfaOptions;
  }

  const mfaEnabled = mfa === 'on' || mfa === 'optional';
  const preferMfaWebauthn = mfaEnabled && preferredMfa === 'webauthn';
  const preferMfaU2f = mfaEnabled && preferredMfa === 'u2f';

  if (mfa === 'webauthn' || preferMfaWebauthn) {
    mfaOptions.push({ value: 'webauthn', label: 'Hardware Key' });
  }

  if (mfa === 'u2f' || preferMfaU2f) {
    mfaOptions.push({ value: 'u2f', label: 'Hardware Key' });
  }

  if (mfa === 'otp' || mfaEnabled) {
    mfaOptions.push({ value: 'otp', label: 'Authenticator App' });
  }

  if (!requireMfa && mfa === 'optional') {
    mfaOptions.push({ value: 'optional', label: 'None' });
  }

  return mfaOptions;
}
