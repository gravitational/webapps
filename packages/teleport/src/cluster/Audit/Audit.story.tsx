/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { Router } from 'react-router';
import { createMemoryHistory } from 'history';
import TeleportContextProvider from 'teleport/teleportContextProvider';
import TeleportContext from 'teleport/teleportContext';
import Audit from './Audit';

export default {
  title: 'TeleportCluster/Audit',
};

export const AuditLog = () => {
  const ctx = new TeleportContext();
  ctx.auditService.fetchEvents = () =>
    Promise.resolve({
      overflow: false,
      events: [],
    });

  return render(ctx);
};

export const Overflow = () => {
  const ctx = new TeleportContext();
  ctx.auditService.fetchEvents = () =>
    Promise.resolve({
      overflow: true,
      events: [],
    });

  return render(ctx);
};

export const Processing = () => {
  const ctx = new TeleportContext();
  ctx.auditService.fetchEvents = () => new Promise(() => null);
  return render(ctx);
};

export const Failed = () => {
  const ctx = new TeleportContext();
  ctx.auditService.fetchEvents = () =>
    Promise.reject(new Error('server error'));
  return render(ctx);
};

function render(ctx) {
  const history = createMemoryHistory({
    initialEntries: ['/web/cluster/localhost/audit/events'],
    initialIndex: 0,
  });

  return (
    <TeleportContextProvider value={ctx}>
      <Router history={history}>
        <Audit />
      </Router>
    </TeleportContextProvider>
  );
}
