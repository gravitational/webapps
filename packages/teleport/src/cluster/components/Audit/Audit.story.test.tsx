import React from 'react';
import { Overflow } from './Audit.story';
import { render } from 'design/utils/testing';

test('overflow', async () => {
  const { container, findByText } = render(<Overflow />);
  await findByText(/exceeded the maximum limit of 9999 events/);

  expect(container.firstChild).toMatchSnapshot();
});
