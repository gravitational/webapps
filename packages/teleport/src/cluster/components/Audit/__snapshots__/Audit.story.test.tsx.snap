// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overflow 1`] = `
.c6 {
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 2px;
  box-sizing: border-box;
  box-shadow: 0 1px 4px rgba(0,0,0,.24);
  margin: 0 0 24px 0;
  min-height: 40px;
  padding: 8px 16px;
  overflow: auto;
  word-break: break-all;
  line-height: 1.5;
  background: #f50057;
  color: #FFFFFF;
}

.c6 a {
  color: #FFFFFF;
}

.c5 {
  display: inline-block;
  transition: color .3s;
  margin-left: 8px;
  margin-right: -8px;
  color: rgba(255,255,255,0.87);
  font-size: 14px;
}

.c12 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
  font-size: 16px;
}

.c14 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
}

.c16 {
  display: inline-block;
  transition: color .3s;
  margin-right: 8px;
  color: #FFFFFF;
}

.c3 {
  line-height: 1.5;
  border-radius: 4px;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  border: 1px solid;
  box-sizing: border-box;
  background-color: transparent;
  cursor: pointer;
  font-family: inherit;
  font-weight: bold;
  outline: none;
  opacity: .56;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all .3s;
  -webkit-font-smoothing: antialiased;
  border-color: rgba(255,255,255,0.87);
  color: #FFFFFF;
  font-size: 10px;
  min-height: 24px;
  padding: 0px 16px;
  margin-left: auto;
  width: auto;
  border-color: rgba(255,255,255,0.87);
  color: rgba(255,255,255,0.87);
}

.c3:hover {
  opacity: 1;
}

.c3:active {
  opacity: .24;
}

.c3 > span {
  display: flex;
  align-items: center;
  justify-content: center;
}

.c3:hover,
.c3:focus {
  border-color: #FFFFFF;
  color: #FFFFFF;
}

.c3:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c3:hover,
.c3:focus {
  border-color: #FFFFFF;
  color: #FFFFFF;
}

.c10 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 12px;
  line-height: 16px;
  margin: 0px;
  color: #FFFFFF;
}

.c15 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 16px;
  line-height: 32px;
  margin: 24px;
  color: rgba(255,255,255,0.87);
}

.c7 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  justify-content: flex-start;
}

.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  margin-right: 56px;
  flex-shrink: 0;
}

.c0 {
  box-sizing: border-box;
  padding-left: 40px;
  padding-right: 40px;
  display: flex;
  overflow: auto;
  width: 100%;
  height: 100%;
  flex-direction: column;
}

.c0::after {
  content: ' ';
  padding-bottom: 24px;
}

.c13 {
  background: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-collapse: collapse;
  border-spacing: 0;
  border-radius: 8px;
  font-size: 12px;
  width: 100%;
}

.c13 > thead > tr > th,
.c13 > tbody > tr > th,
.c13 > tfoot > tr > th,
.c13 > thead > tr > td,
.c13 > tbody > tr > td,
.c13 > tfoot > tr > td {
  padding: 8px 8px;
  vertical-align: middle;
}

.c13 > thead > tr > th:first-child,
.c13 > tbody > tr > th:first-child,
.c13 > tfoot > tr > th:first-child,
.c13 > thead > tr > td:first-child,
.c13 > tbody > tr > td:first-child,
.c13 > tfoot > tr > td:first-child {
  padding-left: 24px;
}

.c13 > thead > tr > th:last-child,
.c13 > tbody > tr > th:last-child,
.c13 > tfoot > tr > th:last-child,
.c13 > thead > tr > td:last-child,
.c13 > tbody > tr > td:last-child,
.c13 > tfoot > tr > td:last-child {
  padding-right: 24px;
}

.c13 > thead > tr > th {
  background: #111B48;
  color: #FFFFFF;
  cursor: pointer;
  font-size: 10px;
  font-weight: 400;
  padding-bottom: 0;
  padding-top: 0;
  text-align: left;
  opacity: 0.75;
  text-transform: uppercase;
  white-space: nowrap;
}

.c13 > thead > tr > th .c4 {
  font-weight: bold;
  font-size: 8px;
  margin-left: 8px;
}

.c13 > tbody > tr > td {
  color: rgba(255,255,255,0.87);
  line-height: 16px;
}

.c13 tbody tr {
  border-bottom: 1px solid #1C254D;
}

.c13 tbody tr:hover {
  background-color: rgb(37,49,98);
}

.c11 {
  box-sizing: border-box;
  display: flex;
}

.c11 button {
  background: none;
  border: none;
  border-radius: 200px;
  cursor: pointer;
  padding: 0;
  margin: 0 0 0 8px;
  outline: none;
  transition: all 0.3s;
  text-align: center;
}

.c11 button:hover,
.c11 button:focus {
  background: #1C254D;
}

.c11 button:hover .c4,
.c11 button:focus .c4 {
  opacity: 1;
}

.c11 button .c4 {
  opacity: 0.4;
  font-size: 20px;
  transition: all 0.3s;
}

.c9 {
  padding: 8px 24px;
  display: flex;
  height: 24px;
  flex-shrink: 0;
  align-items: center;
  justify-content: space-between;
  background: #222C59;
  border-top-left-radius: 8px;
  border-top-right-radius: 8px;
}

.c8 {
  box-sizing: border-box;
  font-size: 12px;
  min-width: 200px;
  outline: none;
  border: none;
  border-radius: 200px;
  height: 32px;
  transition: all .2s;
  background: #222C59;
  margin-right: 16px;
  padding-left: 16px;
  padding-right: 16px;
  color: rgba(255,255,255,0.87);
  height: 30px;
}

.c8:hover {
  background: #2C3A73;
}

.c8:focus,
.c8:active {
  background: #2C3A73;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  color: rgba(255,255,255,0.87);
}

.c8::placeholder {
  color: rgba(255,255,255,0.24);
  font-size: 12px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      Audit Log
    </div>
    <button
      class="c3"
      style="min-width: 100px;"
      width="auto"
    >
      <span>
        Today
        <span
          class="c4 c5 icon icon-caret-down "
          color="text.primary"
          font-size="2"
        />
      </span>
    </button>
  </div>
  <div
    class="c6"
    kind="danger"
  >
    Number of events retrieved for specified date range has exceeded the maximum limit of 
    9999
     events
  </div>
  <div
    class="c7"
  >
    <input
      class="c8"
      color="text.primary"
      height="30px"
      placeholder="SEARCH..."
      value=""
    />
  </div>
  <div>
    <nav
      class="c9"
    >
      <div
        class="c10"
        color="primary.contrastText"
      >
        SHOWING 
        <strong>
          0
        </strong>
         - 
        <strong>
          0
        </strong>
         
        of 
        <strong>
          0
        </strong>
      </div>
      <div
        class="c11"
      >
        <button
          disabled=""
          title="Previous Page"
        >
          <span
            class="c4 c12 icon icon-arrow-left-circle "
            color="light"
            font-size="3"
          />
        </button>
        <button
          disabled=""
          title="Next Page"
        >
          <span
            class="c4 c12 icon icon-arrow-right-circle "
            color="light"
            font-size="3"
          />
        </button>
      </div>
    </nav>
    <table
      class="c13"
    >
      <thead>
        <tr>
          <th>
            <a>
              Type
              <span
                class="c4 c14 icon icon-chevrons-expand-vertical "
                color="light"
              />
            </a>
          </th>
          <th>
            Description
          </th>
          <th>
            <a>
              Created
              <span
                class="c4 c14 icon icon-chevron-up "
                color="light"
              />
            </a>
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            colspan="4"
          >
            <div
              class="c15"
              color="text.primary"
              style="display: flex; align-items: center; justify-content: center;"
            >
              <span
                class="c4 c16 icon icon-database "
                color="light"
              />
              No Data Available
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
